source/opt/workaround_image_operands.cpp

```diff
@@ -0,0 +1,160 @@
+
+        auto new_sampled_image = op_sampled_image->Clone(context());
+        new_sampled_image->SetResultType(sampled_image_type);
+
+        new_sampled_image->InsertBefore(op_sampled_image);
+        new_sampled_image->SetResultId(context()->TakeNextId());
+
+        sampled_image_usage->SetOperand(idx, {new_sampled_image->result_id()});
+        context()->AnalyzeUses(sampled_image_usage);
+        context()->AnalyzeDefUse(new_sampled_image);
+      });
+}
+
+void WorkaroundImageOperands::WalkImageUse(
+    Instruction* inst, Instruction* img_type,
+    analysis::DefUseManager* def_use_mgr) {
+  analysis::SampledImage* sampled_image_type =
+      context()->get_type_mgr()->GetType(inst->result_id())->AsSampledImage();
+
+  context()->get_def_use_mgr()->ForEachUse(
+      inst, [this, img_type, sampled_image_type, def_use_mgr](
+                opt::Instruction* op_type_sampled_image_usage, uint32_t idx) {
+        if (op_type_sampled_image_usage->opcode() == SpvOpSampledImage) {
+          WalkSampledImageUse(op_type_sampled_image_usage, idx, img_type,
+                              sampled_image_type, def_use_mgr);
+        }
+      });
+}
+
+bool WorkaroundImageOperands::FixupOpTypeImage() {
+  bool modified = false;
+  auto* module = get_module();
+  auto* def_use_mgr = context()->get_def_use_mgr();
+  std::vector<opt::Instruction*> unknown_depth_images;
+
+  for (Module::inst_iterator type_it = context()->types_values_begin();
+       type_it != context()->types_values_end(); ++type_it) {
+    // 2 means no indication as to whether this is a depth or non-depth image
+    if (type_it->opcode() == SpvOpTypeImage) {
+      if (type_it->GetSingleWordOperand(3) == 2) {
+        unknown_depth_images.push_back(&(*type_it));
+        modified = true;
+      }
+    }
+  }
+
+  std::vector<opt::Instruction*> fixed_instructions;
+
+  for (auto img_type : unknown_depth_images) {
+    def_use_mgr->ForEachUse(
+        img_type->result_id(),
+        [this, img_type, def_use_mgr](opt::Instruction* inst, uint32_t) {
+          if (inst->opcode() == SpvOpTypeSampledImage) {
+            WalkImageUse(inst, img_type, def_use_mgr);
+          }
+        });
+  }
+
+  for (auto img_type : unknown_depth_images) {
+    analysis::Image* image_type =
+        context()->get_type_mgr()->GetType(img_type->result_id())->AsImage();
+    analysis::Type* t = context()->get_type_mgr()->GetType(
+        context()->get_type_mgr()->GetId(image_type->sampled_type()));
+    analysis::Image img(t, image_type->dim(), 0, image_type->is_arrayed(),
+                        image_type->is_multisampled(), image_type->sampled(),
+                        image_type->format(), image_type->access_qualifier());
+    for (auto& dec : image_type->decorations()) {
+      img.AddDecoration(std::vector<uint32_t>(dec));
+    }
+    uint32_t new_image = context()->get_type_mgr()->GetTypeInstruction(&img);
+    context()->ReplaceAllUsesWith(img_type->result_id(), new_image);
```

Comment:
For this change, as you mentioned, it is possible that `new_image` does not dominate all of the uses of `img_type`, so using`ReplaceAllUsesWith` can create invalid code.

This can happen two ways:  `new_image` is a newly created type or it is an existing type that was declared after `img_type`.  I see two possible solution.  The first it to try to rewrite the `img_type`.  You can only do this if you are sure there it will not create a duplicate type.  

The other would be to move `new_image` to just before `img_type` (`new_image->InsertAfter(img_type)`).  You would have to do a scan to make sure `new_image` is being moved earlier and not later.